### PR TITLE
Upgrade karpenter to 0.13.1

### PR DESCRIPTION
**Pull request by rifelpet**

### CODE REVIEW

**Comment by rifelpet** [comment]
I think we need a way to extend this duration for e2e tests: https://github.com/kubernetes/kops/blob/31799bf81234656340c9e57661893960aac44da4/upup/models/cloudup/resources/addons/karpenter.sh/k8s-1.19.yaml.template#L850

https://karpenter.sh/v0.13.1/provisioner/#specttlsecondsafterempty

The [karpenter-controller logs](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kops/13918/pull-kops-e2e-aws-karpenter/1542274893779832832/artifacts/cluster-info/kube-system/karpenter-controller-6568cc86d6-g7c2g/logs.txt) mention 17 occurrences of `Triggering termination after 30s for empty node` and I wonder if the volume test timeouts we're seeing are because there's too much node churn in the cluster.

Any ideas how we could conditionally increase that setting?

**Comment by olemarkus** [comment]
> I think we need a way to extend this duration for e2e tests:
> 
> https://github.com/kubernetes/kops/blob/31799bf81234656340c9e57661893960aac44da4/upup/models/cloudup/resources/addons/karpenter.sh/k8s-1.19.yaml.template#L850
> 
> https://karpenter.sh/v0.13.1/provisioner/#specttlsecondsafterempty
> 
> The [karpenter-controller logs](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kops/13918/pull-kops-e2e-aws-karpenter/1542274893779832832/artifacts/cluster-info/kube-system/karpenter-controller-6568cc86d6-g7c2g/logs.txt) mention 17 occurrences of `Triggering termination after 30s for empty node` and I wonder if the volume test timeouts we're seeing are because there's too much node churn in the cluster.
> 
> Any ideas how we could conditionally increase that setting?

I suggest mapping the field from https://github.com/kubernetes/kops/pull/13892 to this parameter.

The current value was intentionally set low to see how karpenter deals with volatility. In production this value probably should be whatever the value for "infinite" is. Karpenter on EKS sort of require this value as a poor mans upgrade job/bin packing. But since kOps is aware of these instances and can do a regular rolling update, this feature is less useful. At least when karpenter also starts supporting condensing Pods.

File: upup/models/cloudup/resources/addons/karpenter.sh/k8s-1.19.yaml.template [hunk 1]
```diff
@@ -676,6 +714,17 @@ webhooks:
...
   rules:
+  - apiGroups:
+    - karpenter.k8s.aws
+    apiVersions:
+    - v1alpha1
+    operations:
+    - CREATE
+    - UPDATE
+    resources:
+    - awsnodetemplates
```
Comment (olemarkus) [inline upup/models/cloudup/resources/addons/karpenter.sh/k8s-1.19.yaml.template:725]: Are these CRDs managed by karpenter itself? I don't see that we are applying them to the cluster.

**Comment by hakman** [comment]
Can we just set `ttlSecondsAfterEmpty: 300`?

**Comment by olemarkus** [comment]
> Can we just set `ttlSecondsAfterEmpty: 300`?

Yeah. I confused this with `ttlSecondsUntilExpired`

File: upup/models/cloudup/resources/addons/karpenter.sh/k8s-1.19.yaml.template [hunk 2]
```diff
@@ -282,33 +282,38 @@ kind: ClusterRole
...
-  resources: ["storageclasses"]
+- apiGroups: ["karpenter.k8s.aws"]
+  resources: ["awsnodetemplates"]
   verbs: ["get", "list", "watch"]
 - apiGroups: [""]
-  resources: ["nodes", "pods"]
-  verbs: ["get", "list", "watch", "patch", "delete"]
-- apiGroups: [""]
-  resources: ["configmaps"]
+  resources: ["pods", "nodes", "persistentvolumes", "persistentvolumeclaims"]
```
Comment (hakman) [inline upup/models/cloudup/resources/addons/karpenter.sh/k8s-1.19.yaml.template:293]:
```
E0629 23:39:29.187577       1 reflector.go:138] k8s.io/client-go@v0.21.4/tools/cache/reflector.go:167: Failed to watch *v1.Namespace:
failed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:kube-system:karpenter" cannot list resource "namespaces" in API group "" at the cluster scope
```
```suggestion
  resources: ["pods", "nodes", "namespaces", "persistentvolumes", "persistentvolumeclaims"]
```

**Review by hakman — CHANGES_REQUESTED** [review]

**Comment by olemarkus** [comment]
Just skip the in-tree tests?

**Comment by rifelpet** [comment]
I agree we should skip the in-tree tests

**Comment by rifelpet** [comment]
We're now down to just the node ssh test, I'm not sure why it can't find any external IPs even though there is one on every node:

```
 [BeforeEach] [sig-node] SSH
  test/e2e/node/ssh.go:38
[It] should SSH to all nodes and run commands
  test/e2e/node/ssh.go:47
STEP: Getting all nodes' SSH-able IP addresses
Jul  2 23:34:15.230: INFO: No external IP address on nodes, falling back to internal IPs
Jul  2 23:34:18.230: INFO: cannot dial 172.20.37.87:22: dial tcp 172.20.37.87:22: i/o timeout
Jul  2 23:34:18.230: INFO: Skipping host 172.20.37.87 because it does not run anything on port 22
Jul  2 23:34:18.230: INFO: cannot dial 172.20.63.174:22: dial tcp 172.20.63.174:22: i/o timeout
Jul  2 23:34:18.230: INFO: Skipping host 172.20.63.174 because it does not run anything on port 22
Jul  2 23:34:18.230: INFO: cannot dial 172.20.57.186:22: dial tcp 172.20.57.186:22: i/o timeout
Jul  2 23:34:18.230: INFO: Skipping host 172.20.57.186 because it does not run anything on port 22
Jul  2 23:34:18.230: INFO: cannot dial 172.20.40.244:22: dial tcp 172.20.40.244:22: i/o timeout
Jul  2 23:34:18.230: INFO: cannot dial 172.20.34.109:22: dial tcp 172.20.34.109:22: i/o timeout
Jul  2 23:34:18.230: INFO: cannot dial 172.20.49.4:22: dial tcp 172.20.49.4:22: i/o timeout
Jul  2 23:34:18.230: INFO: Skipping host 172.20.49.4 because it does not run anything on port 22
Jul  2 23:34:18.230: INFO: Skipping host 172.20.34.109 because it does not run anything on port 22
Jul  2 23:34:18.230: INFO: Skipping host 172.20.40.244 because it does not run anything on port 22
Jul  2 23:34:18.230: INFO: cannot dial 172.20.42.135:22: dial tcp 172.20.42.135:22: i/o timeout
Jul  2 23:34:18.230: INFO: Skipping host 172.20.42.135 because it does not run anything on port 22
Jul  2 23:34:18.230: INFO: cannot dial 172.20.46.233:22: dial tcp 172.20.46.233:22: i/o timeout
Jul  2 23:34:18.230: INFO: Skipping host 172.20.46.233 because it does not run anything on port 22
Jul  2 23:34:18.230: INFO: cannot dial 172.20.50.78:22: dial tcp 172.20.50.78:22: i/o timeout
Jul  2 23:34:18.230: INFO: Skipping host 172.20.50.78 because it does not run anything on port 22
Jul  2 23:34:18.230: INFO: cannot dial 172.20.33.125:22: dial tcp 172.20.33.125:22: i/o timeout
Jul  2 23:34:18.230: INFO: Skipping host 172.20.33.125 because it does not run anything on port 22
Jul  2 23:34:18.230: INFO: cannot dial 172.20.56.177:22: dial tcp 172.20.56.177:22: i/o timeout
Jul  2 23:34:18.230: INFO: Skipping host 172.20.56.177 because it does not run anything on port 22
STEP: Found 0 SSH'able hosts
STEP: SSH'ing to 0 nodes and running echo "Hello from $(whoami)@$(hostname)"
STEP: SSH'ing to 1 nodes and running echo "foo" | grep "bar"
STEP: SSH'ing host 
error dialing ubuntu@: 'dial tcp: missing address', retrying
error dialing ubuntu@: 'dial tcp: missing address', retrying
error dialing ubuntu@: 'dial tcp: missing address', retrying
error dialing ubuntu@: 'dial tcp: missing address', retrying
error dialing ubuntu@: 'dial tcp: missing address', retrying
Jul  2 23:34:38.235: FAIL: Ran echo "foo" | grep "bar" on , got error error getting SSH client to ubuntu@: 'timed out waiting for the condition', expected <nil> 
```

```yaml
  status:
    addresses:
    - address: 172.20.46.233
      type: InternalIP
    - address: 15.236.224.179
      type: ExternalIP
    - address: i-052b809a11c58e5e2.eu-west-3.compute.internal
      type: InternalDNS
    - address: i-052b809a11c58e5e2.eu-west-3.compute.internal
      type: Hostname
    - address: ec2-15-236-224-179.eu-west-3.compute.amazonaws.com
      type: ExternalDNS
```

**Comment by olemarkus** [comment]
I cannot reproduce the SSH error. Not that I believe it is related, but this PR is still missing the CRD? As well as karpenter-global-settings config map?

**Comment by hakman** [comment]
As we can see in https://testgrid.k8s.io/kops-misc#kops-aws-misc-karpenter, SSH test is flaky.

**Comment by rifelpet** [comment]
Looking at the new CRD, is that something we want to support? seems like most settings we would want hardcoded through instance groups so I'm not sure what value there is in allowing users to customize this directly. Maybe I'm wrong though

https://github.com/aws/karpenter/blob/main/charts/karpenter/crds/karpenter.k8s.aws_awsnodetemplates.yaml

**Comment by olemarkus** [comment]
Yeah I don't think we want to support this

**Comment by olemarkus** [comment]
The SSH error .. can it happen because nodes are not properly initialized? I have run this many times locally using this PR branch and I cannot reproduce it

**Comment by rifelpet** [comment]
I suppose if the test were to query the list of nodes at the time between a node resource being created by kubelet and CCM populating the node's addresses, that could lead to this test failing.

I just increased the ttl to 5 minutes (similar to hakman's PR) so we'll see if that helps reduce node churn.

/test pull-kops-e2e-aws-karpenter

File: upup/pkg/fi/cloudup/template_functions.go [hunk 3]
```diff
@@ -327,6 +327,12 @@ func (tf *TemplateFunctions) AddTo(dest template.FuncMap, secretStore fi.SecretS
 	dest["KarpenterInstanceTypes"] = func(ig kops.InstanceGroupSpec) ([]string, error) {
 		return karpenterInstanceTypes(tf.cloud.(awsup.AWSCloud), ig)
 	}
+	dest["KarpenterTTLSecondsAfterEmpty"] = func(ig kops.InstanceGroupSpec) int64 {
+		if ig.MaxInstanceLifetime != nil {
+			return int64(ig.MaxInstanceLifetime.Seconds())
+		}
+		return 300
+	}
```
Comment (hakman) [inline upup/pkg/fi/cloudup/template_functions.go:335]: I don't think this mapping is good. `MaxInstanceLifetime` has nothing to do with this.
We should just set `ttlSecondsAfterEmpty: 300` or `180`.

**Comment by hakman** [comment]
I think the CRDs should be updated even though there are no new settings useful to us. This would prevent possible crashes on the controller side.

File: upup/pkg/fi/cloudup/template_functions.go (hunk 3 above)
Comment (olemarkus) [reply upup/pkg/fi/cloudup/template_functions.go:335]: It was a mistake on my end. I confused this one with `ttlSecondsUntilExpired`

**Comment by rifelpet** [comment]
I've added the new CRD and hardcoded the ttlSecondsAfterEmpty to 300

**Pull request merged**